doc/design/listener.md

```diff
@@ -0,0 +1,306 @@
+| ConditionListener                             | | ConditionNotifier                                |
+|   ConditionListener(ConditionVariableData & ) | |   ConditionNotifier(ConditionVariableData &,     |
+|                                               | |                 uint64_t notificationIndex)      |
+|   bool                  wasNotified()         | |                                                  |
+|   void                  destroy()             | |   void notify()                                  |
+|   NotificationVector_t  wait()                | |                                                  |
+|   NotificationVector_t  timedWait()           | |   - m_condVarDataPtr    : ConditionVariableData* |
+|                                               | |   - m_notificationIndex                          |
+|   - m_condVarDataPtr : ConditionVariableData* | +--------------------------------------------------+
+|   - m_toBeDestroyed  : std::atomic_bool       |        | 1
++-----------------------------------------------+        |
+        | 1                                              | n
+        |                                           +--------------------------------+
+        | 1                                         | TriggerHandle                  |
++-------------------------------------------------+ |   bool isValid()               |
+| Listener                                        | |   bool wasTriggered()          |
+|   attachEvent(Triggerable, EventType, Callback) | |   void trigger()               |
+|   detachEvent(Triggerable, EventType)           | |   void reset()                 |
+|                                                 | |   void invalidate()            |
+|   - m_events : Event_t[]                        | |   void getUniqueId()           |
+|   - m_thread : std::thread                      | |                                |
+|   - m_conditionListener : ConditionListener     | |   - m_conditionVariableDataPtr |
+|                                                 | |   - m_resetCallback            |
+| +----------------------------+                  | |   - m_uniqueTriggerId          |
+| | Event_t                    |                  | +--------------------------------+
+| |   void executeCallback()   |                  |      | 1
+| |   bool reset()             |                  |      |
+| |   bool init(...)           |                  |      | n
+| |                            |                  | +----------------------------------------------------+
+| |   - m_origin               |                  | | Triggerable (e.g. Subscriber)                      |
+| |   - m_callback             |                  | |                                                    |
+| |   - m_invalidationCallback |                  | |   void invalidateTrigger(const uint64_t triggerId) |
+| |   - m_eventId              |                  | |   void enableEvent(TriggerHandle&&)                |
```

Comment:
Doesn't `enableEvent` also require an event argument?

Reply:
In the single event case not necessarily, right? But then `disableEvent` doesn't require any argument, too.